lib/puma/cluster.rb

```diff
@@ -227,11 +212,13 @@ def check_workers(force=false)
       # during this loop by giving the kernel time to kill them.
       sleep 1 if any
 
-      pids = []
-      while pid = Process.waitpid(-1, Process::WNOHANG) do
-        pids << pid
+      @workers.reject! do |w|
```

Comment:
this code now seems similar enough to be extracted into a new method, yes?

Reply:
I did think about that, but see below.

Reply:
This logic seems like the same one that causing zombies because we will delete from `@workers` if a process is dead due to signal but the process is still running. When it actually does exit, we won't call wait on it, so this reintroduces the bug.

Reply:
> if a process is dead due to signal

I did 'turn around' things so that wait was called before `dead?` was checked, but I've kind of left the `dead?` check in the because I thought it was previously used for a long time.  But, `wait` may not have been using `WNOHANG` then.  Also, I need to review how & when `@dead` is set, as I'm not clear on that.

Regardless, what you're saying makes sense to me, should I remove the ` || w.dead?` code on line 220?

Reply:
@evanphx

As above, I removed the `|| w.dead?` code in my fork, and it passed.  Would you prefer that?  I think I would...

I don't think there are any tests that create a situation where a worker needs a 'graceful' shutdown and then externally closes it, and the waits to see if a zombie exists.  Maybe a bit messy...

Reply:
I removed `|| w.dead?`

Reply:
This changes the logic so we'll need to valid it: namely should workers be respawned if they are marked as dead but the process still exists? Prior to this change, it did that. But with this change, we don't respawn dead workers until the process actually goes away. We need to decide which is right.

Reply:
IDN.  If this does cause an issue, maybe there needs to be a  'pid array', which is just used for `wait` calls?  Also remove from workers if `wait` returns the pid?

Reply:
I haven't had time to look at `@dead`.  If a worker pid is killed (SIGTERM) outside of Puma, will `@dead` be true?

Reply:
Using a separate array is what I started to do in that other PR. I don't think it's probably necessary if we remove `w.dead?` from here, but I'm mentioning it just so that we've at least thought about it.

Reply:
I looked through the code and it's fine. The main effect will be that there will be a very small delay perhaps before spawning a replacement worker and that's ok. The dead status code is sent right before the worker exits anyway.

Reply:
Thanks for looking at it.  Next major computer purchase is a system for Ubuntu/*nix something...

Reply:
@MSP-Greg or just rent an AWS instance by the hour!

Reply:
@evanphx 

After finally noticing the `reject!` typo (funny how `reject` doesn't work), I think I've got the code for a 'separate array' working.  If you'd prefer it, let me know.

Patch: https://github.com/MSP-Greg/puma/commit/f141b027f14421df0d45c29af43b6af17c38560d
Travis: https://travis-ci.org/MSP-Greg/puma/builds/569122723

@nateberkopec you're on macOS I assume.  You're 90% of the way there already.  I might as well be in the next solar system...